nym-vpn-lib/src/connection_monitor.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: GPL-3.0-only
+
+use std::time::{Duration, Instant};
+
+use futures::{channel::mpsc, StreamExt};
+use nym_task::TaskClient;
+use tokio::task::JoinHandle;
+use tracing::{debug, error, trace};
+
+use crate::error::Result;
+
+const CONNECTION_MONITOR_REPORT_INTERVAL: Duration = Duration::from_secs(5);
+
+// When the latest successful self ping is older than this, we consider the connection to the
+// entry gateway as down
+const SELF_PING_EXPIRY: Duration = Duration::from_secs(5);
+
+#[derive(Debug)]
+pub enum ConnectionStatusEvent {
+    MixnetSelfPing,
+}
+
+#[derive(Debug, Default)]
+struct ConnectionStats {
+    // TODO: extend with all sorts of good stuff
+    latest_self_ping: Option<Instant>,
+}
+
+struct ConnectionMonitor {
+    connection_event_rx: mpsc::UnboundedReceiver<ConnectionStatusEvent>,
+    stats: ConnectionStats,
+}
+
+impl ConnectionMonitor {
+    fn new(connection_event_rx: mpsc::UnboundedReceiver<ConnectionStatusEvent>) -> Self {
+        ConnectionMonitor {
+            connection_event_rx,
+            stats: ConnectionStats::default(),
+        }
+    }
+
+    async fn run(mut self, mut task_client: TaskClient) -> Result<()> {
+        debug!("Connection monitor is running");
+        let mut report_interval = tokio::time::interval(CONNECTION_MONITOR_REPORT_INTERVAL);
+        // Reset so that we don't send a report immediately before we even have a change for any
+        // self pings to be sent and received
+        report_interval.reset();
+
+        loop {
+            tokio::select! {
+                _ = task_client.recv() => {
+                    trace!("ConnectionMonitor: Received shutdown");
+                    break;
+                }
+                Some(event) = self.connection_event_rx.next() => {
+                    match event {
+                        ConnectionStatusEvent::MixnetSelfPing => {
+                            trace!("Received self ping event");
+                            self.stats.latest_self_ping = Some(Instant::now());
+                        }
+                    }
+                }
+                _ = report_interval.tick() => {
+                    debug!("Connection statistics: {:?}", self.stats);
+
+                    // Send hello messages, so listerners can hear that we are still alive
+                    task_client.send_status_msg(Box::new(ConnectionMonitorStatus::Hello));
+
+                    if let Some(latest_self_ping) = self.stats.latest_self_ping {
+                        if latest_self_ping.elapsed() > SELF_PING_EXPIRY {
```

Comment:
Suggestion: Maybe use the opposite of this affirmation and the `continue` action, and the print and error message outside of the outer `if` statement, so that you don't duplicate it

Reply:
Yep it's a imho a good idea to dedup, although in this case I expect these two branches to diverge very soon with additional stats put in place